Adding hook feature

intru91 opened the pull request:
to execute external scripts when special events occur.

It can be use to execute pandoc or asciidoc after editing a note for example.
This first version has only post hook script which is executed after the editor closing

Commit:
Adding hook feature to execute external scripts when special events
occure.

This first version has only post hook script which is executed after the
editor closing

seanh commented:
Hi @intru91 , this sounds awesome. A hook feature would also be good for calling a vcs like git to version your notes. I'll get round to merging this asap (at the weekend, or next week)

seanh commented on bin/terminal_velocity:
```diff
@@ -60,6 +61,11 @@ the default default will be used"""
 
+    parser.add_argument("--hooks", dest="hooks", action="store",
+        default=defaults.get("hooks", "{}"),
+        help="the scripts which get executed when events occure such has \
```
Typos:

occure -> occur
such has -> such as

Also I don't think the `\` is necessary because Python has implied line continuation inside parentheses, so you can do:

```
parser.add_argument(...
    help="the scripts which get executed when events occur such as "
         "saving notes (default: %(default)s)")
```

intru91 replied:
I get an error on python 2.7.9 when trying:
`print("test
test")
`

seanh commented on terminal_velocity/urwid_ui.py:
```diff
@@ -306,6 +308,15 @@ def quit(self):
+        For the time being only the post hook has been tested
+
+        """
+        if hook in self.hooks != "" and os.path.isfile(self.hooks[hook]):
```
This is a bit odd. `hook in self.hooks` is a boolean expression, it will always equal `True` or `False`, never `""`, so `if hook in self.hooks != ""` will always be `True`.

I think this might do what you wanted: `if self.hooks.get(hook) and ...`.

seanh commented:
Thanks @intru91, I added a couple of small things that I'd like you to fix.

There's also a couple of slightly bigger things:

Would you mind changing the approach slightly? Rather than a `--hooks` argument that takes JSON, I'd prefer a `--post-hook` argument that just takes a simple command to execute as argument (the command could be a path to a script file or just a command). This avoids bringing JSON into it, so keeps it simpler.

I think it'd also be nice to add some documentation (in the help text for the `--post-hook` argument maybe) explaining what the post hook is, i.e. that it gets executed after leaving the editor and returning to tv (and it looks like it always gets executed, even if the file wasn't edited, which is fine but worth a mention in the help text).

intru91 commented:
Hi @seanh, thank you for the feedback !

I'll fix the typos.

For `if hook in self.hooks != ""`, nice catch, it was a old code i left hanging ... . I'll replace it by `if hook in self.hooks and ...` 

As for the user interface, JSON was a convenient way of getting data in the internal of terminal_velocity but was really bad for the user (hard to get right from the shell and exposing the internal data structures....). I actually intended to change the interface to a similar one that you described. I'll work on it and give you a feedback when i'll have something satisfying.

seanh commented:
Thanks! Looking forward to it

Commit:
Replacing JSON parameter style by a --*-hook parameter style

    For example post hook is now --post-hook on the command line.
    Internally, hooks are still represented as a dict.

Commit:
Adding try except when launching hook.

intru91 commented:
Hi @seanh,

I'm pushed two new commit on my branch to apply the changes that you suggested.

seanh commented:
@intru91 This looks great. It hard merge conflicts with the current master branch. Can you either pull master into your branch or rebase your branch on master (whichever you prefer) and solve the merge conflicts?

Commit:
Merge branch 'master' of https://github.com/seanh/terminal_velocity

Conflicts:
	bin/terminal_velocity
	terminal_velocity/urwid_ui.py
